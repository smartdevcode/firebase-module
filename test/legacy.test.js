const { resolve } = require('path')
const fs = require('fs-extra')
const { Nuxt, Builder } = require('nuxt')
const { injectionMapping } = require('../lib/template-utils')
const FirebaseModule = require('..')

jest.mock('firebase/app', () => ({
  apps: [
    new Proxy({}, {
      get (target, name) {
        return jest.fn(() => ({}))
      }
    })
  ],
  messaging: {
    isSupported: jest.fn()
  }
}))

describe('legacy', () => {
  let nuxt
  const buildDir = resolve(__dirname, '.nuxt-legacy')

  beforeAll(async () => {
    const config = {
      rootDir: resolve(__dirname, '..'),
      buildDir,
      srcDir: resolve(__dirname, 'fixture'),
      modules: [
        [FirebaseModule, {
          legacyMode: true,
          config: {
            // REQUIRED: Official config for firebase.initializeApp(config):
            apiKey: '<apiKey>',
            authDomain: '<authDomain>',
            databaseURL: '<databaseURL>',
            projectId: '<projectId>',
            storageBucket: '<storageBucket>',
            messagingSenderId: '<messagingSenderId>',
            appId: '<appId>',
            measurementId: '<measurementId>'
          },
          services: {
            analytics: true,
            auth: true,
            firestore: true,
            functions: true,
            messaging: true,
            performance: true,
            realtimeDb: true,
            remoteConfig: true,
            storage: true
          }
        }]
      ]
    }
    config.dev = false

    nuxt = new Nuxt(config)
    const BundleBuilder = { build: _ => _ }
    const builder = new Builder(nuxt, BundleBuilder)
    await builder.build()
  }, 60000)

  afterAll(async () => {
    await nuxt.close()
  })

  test('plugin contents', async () => {
    const content = await fs.readFile(resolve(buildDir, 'firebase/index.js'), { encoding: 'utf8' })

    for (const injectedName of Object.values(injectionMapping)) {
      expect(content).toContain(`'fire${injectedName}'`)
      expect(content).toContain(`'fire${injectedName}Obj'`)
    }

    expect(content).not.toContain('fire.auth')

    expect(content).toMatchSnapshot()
  })

  test('exec plugin (server)', async () => {
    const Plugin = await import(resolve(buildDir, 'firebase/index.js')).then(m => m.default || m)
    const inject = jest.fn()

    await Plugin({}, inject)
    expect(inject).toHaveBeenCalledTimes(10)
  })

  test('exec plugin (client)', async () => {
    process.client = true
    const Plugin = await import(resolve(buildDir, 'firebase/index.js')).then(m => m.default || m)
    const inject = jest.fn()

    await Plugin({}, inject)
    expect(inject).toHaveBeenCalledTimes(18)
  })
})
