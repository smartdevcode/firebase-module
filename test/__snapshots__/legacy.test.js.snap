// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`legacy plugin contents 1`] = `
"import createApp from './app.js'

import authService from './service.auth.js'
import realtimeDbService from './service.realtimeDb.js'
import firestoreService from './service.firestore.js'
import storageService from './service.storage.js'
import functionsService from './service.functions.js'
import messagingService from './service.messaging.js'
import performanceService from './service.performance.js'
import analyticsService from './service.analytics.js'
import remoteConfigService from './service.remoteConfig.js'

const appConfig = {\\"apiKey\\":\\"\\\\u003CapiKey\\\\u003E\\",\\"authDomain\\":\\"\\\\u003CauthDomain\\\\u003E\\",\\"databaseURL\\":\\"\\\\u003CdatabaseURL\\\\u003E\\",\\"projectId\\":\\"\\\\u003CprojectId\\\\u003E\\",\\"storageBucket\\":\\"\\\\u003CstorageBucket\\\\u003E\\",\\"messagingSenderId\\":\\"\\\\u003CmessagingSenderId\\\\u003E\\",\\"appId\\":\\"\\\\u003CappId\\\\u003E\\",\\"measurementId\\":\\"\\\\u003CmeasurementId\\\\u003E\\"}

export default async (ctx, inject) => {
  const { firebase, session } = await createApp(appConfig)

  let servicePromises = []

  if (process.server) {
    servicePromises = [
      authService(session, firebase, ctx, inject),
    realtimeDbService(session, firebase, ctx, inject),
    firestoreService(session, firebase, ctx, inject),
    storageService(session, firebase, ctx, inject),
    functionsService(session, firebase, ctx, inject),

    ]
  }

  if (process.client) {
    servicePromises = [
      authService(session, firebase, ctx, inject),
      realtimeDbService(session, firebase, ctx, inject),
      firestoreService(session, firebase, ctx, inject),
      storageService(session, firebase, ctx, inject),
      functionsService(session, firebase, ctx, inject),
      messagingService(session, firebase, ctx, inject),
      performanceService(session, firebase, ctx, inject),
      analyticsService(session, firebase, ctx, inject),
      remoteConfigService(session, firebase, ctx, inject),

    ]
  }

  const [
    auth,
    realtimeDb,
    firestore,
    storage,
    functions,
    messaging,
    performance,
    analytics,
    remoteConfig
  ] = await Promise.all(servicePromises)

  if (process.server) {
    inject('fireAuth', auth)

    inject('fireAuthObj', firebase.auth)

    inject('fireDb', realtimeDb)

    inject('fireDbObj', firebase.database)

    inject('fireStore', firestore)

    inject('fireStoreObj', firebase.firestore)

    inject('fireStorage', storage)

    inject('fireStorageObj', firebase.storage)

    inject('fireFunc', functions)

    inject('fireFuncObj', firebase.functions)
  }

  if (process.client) {
    inject('fireAuth', auth)

    inject('fireAuthObj', firebase.auth)

    inject('fireDb', realtimeDb)

    inject('fireDbObj', firebase.database)

    inject('fireStore', firestore)

    inject('fireStoreObj', firebase.firestore)

    inject('fireStorage', storage)

    inject('fireStorageObj', firebase.storage)

    inject('fireFunc', functions)

    inject('fireFuncObj', firebase.functions)

    inject('fireMess', messaging)

    inject('fireMessObj', firebase.messaging)

    inject('firePerf', performance)

    inject('firePerfObj', firebase.performance)

    inject('fireAnalytics', analytics)

    inject('fireAnalyticsObj', firebase.analytics)

    inject('fireConfig', remoteConfig)

    inject('fireConfigObj', firebase.remoteConfig)
  }
}
"
`;
